bloom/bloom.py

```diff
@@ -205,22 +207,41 @@ def _bit_offsets(self, value):
         was *probably* inserted into our Bloom filter.
         '''
         encoded_value = json.dumps(value, sort_keys=True)
-        for seed in range(self.num_hashes()):
+        for seed in xrange(self.num_hashes()):
             yield mmh3.hash(encoded_value, seed=seed) % self.size()
 
     def _load_bit_array(self):
-        bit_string = self.memcache.get(self.key)
+        bit_string, self._cas = self.memcache.gets(self.key)
         if bit_string is None:
             self._bit_array = bitarray('0' * self.size())
+            self._store_bit_array()
         else:
             self._bit_array = bitarray()
             self._bit_array.frombytes(bit_string)
 
     def _store_bit_array(self):
-        if self._bit_array.any():
-            self.memcache.set(self.key, self._bit_array.tobytes())
+        bit_string = self._bit_array.tobytes()
+        if self._cas is None:
+            self.memcache.set(self.key, bit_string)
         else:
-            self.memcache.delete(self.key)
+            if not self.memcache.cas(self.key, bit_string, self._cas):
+                raise CheckAndSetError(memcache=self.memcache, key=self.key)
```

Comment:
Will this only fail if it's a check and set error specifically?  I'm assuming something like a connection error would raise an exception rather than give a false value?  I think retrying instantly is fine in the case of lock contention, but I just don't want there to be a retry storm by every single client if an instance is getting overwhelmed or is unreachable.

Reply:
Excellent point!  I just did some testing and confirmed this:

    >>> from bloom import BloomFilter
    >>> recently_consumed = BloomFilter(key='recently-consumed')
    >>> # Kill the Memcached service...
    ...
    >>> recently_consumed.add('t3_1')
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "bloom/bloom.py", line 255, in add
        self.update({value})
      File "bloom/bloom.py", line 237, in wrapper
        return func(self, *args, **kwargs)
      File "bloom/bloom.py", line 272, in update
        self._store_bit_array()
      File "bloom/bloom.py", line 227, in _store_bit_array
        if not self.memcache.cas(self.key, bit_string, self._cas):
      File "/Users/raj/Documents/Code/reddit/bloom-filter/venv/lib/python2.7/site-packages/pymemcache/client/base.py", line 419, in cas
        return self._store_cmd(b'cas', key, expire, noreply, value, cas)
      File "/Users/raj/Documents/Code/reddit/bloom-filter/venv/lib/python2.7/site-packages/pymemcache/client/base.py", line 800, in _store_cmd
        buf, line = _readline(self.sock, buf)
      File "/Users/raj/Documents/Code/reddit/bloom-filter/venv/lib/python2.7/site-packages/pymemcache/client/base.py", line 1101, in _readline
        raise MemcacheUnexpectedCloseError()
    pymemcache.exceptions.MemcacheUnexpectedCloseError

Also, according to the [pymemcache docs](http://pymemcache.readthedocs.io/en/latest/apidoc/pymemcache.client.base.html?highlight=cas#pymemcache.client.base.Client.cas), the `cas` command returns `True` on success, `False` if the `cas` value has changed (in which case, the value won't have been updated), or `None` if the key doesn't exist.

In other words, we only retry immediately in the case of lock contention.  But if there's a network issue or some other problem with the Memcached service, then we get some other pymemcache exception which we don't catch in our retry decorator and bubble upstream.

Does this seem like a reasonable approach, 💇 @dellis23?

Reply:
Yup, looks good!